### PR TITLE
Dynamic Multi LoRA Load \ Delete Support 

### LINKED ISSUE #3308: Multi-LoRA - Support for providing /load and /unload API
### Problem statement:

In the production system, there should be an API to add\\\\remove fine-tuned weights dynamically. Inference caller should not have to specify LoRA location with each call.

Current Multi-LoRA support allows adaptor load during inference calls, which doesn't check if finetune weights are already loaded and ready for inferencing.

### Proposal:

Introduce an API - /load and /unload to allow fine-tuned weights inclusions in vllm.

`POST /load` -> add finetunes weight as part of models.
`POST /unload` -> remove finetunes weight from models list.

This will allow the set of finetuned weights present in vllm server.

This will infer no need to specify finetune weight names, and locations as part of each inference request.

Sample code:

```python
lora_request = None
index = 1
 
 
@app.post("/load")
async def load(request: Request) -> Response:
    request_dict = await request.json()
    global lora_request
 
    lora_local_path = request_dict.pop("lora_path", "/models/lora/")
    global index
    lora_request = LoRARequest(
        lora_name=lora_local_path,
        lora_int_id=index,
        lora_local_path=lora_local_path)
 
    index = index + 1
    return Response(status_code=201)
 
@app.post("/unload")
async def unload(request: Request) -> Response:
    """
    Unload API
    :param request:
    :return:
    """
    global lora_request
    lora_request = None
 
    global index
    if not index <= 1:
        index = index - 1
 
    return Response(status_code=201)
```

### PR DESCRIPTION
**Pull request by gauravkr2108**
FILL IN THE PR DESCRIPTION HERE

FIX #3308 

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/entrypoints/openai/serving_engine.py [hunk 1]
```diff
@@ -84,6 +84,38 @@ async def show_available_models(self) -> ModelList:
...
+        last_lora_request: LoRARequest = self.lora_requests.get(self.lora_requests.size() - 1)
+        lora_idx = 1
+        if last_lora_request is not None:
+            lora_idx = last_lora_request.lora_int_id + 1
+
+        self.lora_requests.append(LoRARequest(
+            lora_name=lora.name,
+            lora_int_id=lora_idx,
+            lora_local_path=lora.local_path,
+        ))
```
Comment (Yard1) [inline vllm/entrypoints/openai/serving_engine.py:98]: let's have an increment only counter for LoRA ids instead?

also, is it possible to check if the lora has already been added before? we can use the local_path

File: vllm/entrypoints/openai/serving_engine.py [hunk 2]
```diff
@@ -84,6 +84,38 @@ async def show_available_models(self) -> ModelList:
...
+            raise ValueError("Unsupported delete operation, base model delete not supported")
+
+        lora_idx = -1
+        for idx, lora in enumerate(self.lora_requests):
+            if model == lora.lora_name:
+                lora_idx = idx
+        if lora_idx != -1:
+            lora = self.lora_requests.pop(lora_idx)
+            return DeleteResponse(id=lora.lora_name,
+                                  deleted=True)
```
Comment (Yard1) [inline vllm/entrypoints/openai/serving_engine.py:115]: see above - incrementing and decrementing the counter is brittle

**Review by Yard1 — COMMENTED** [review]
Two comments, will also need tests.

**Comment by simon-mo** [comment]
Hi @gauravkr2108, sorry about the delay, please address @Yard1's comment.

File: vllm/entrypoints/openai/serving_engine.py (hunk 1 above)
Comment (gauravkr2108) [reply vllm/entrypoints/openai/serving_engine.py:98]: sorry for the delay, I will be active on this now; I will resume work here.

**Comment by Jeffwan** [comment]
@simon-mo @Yard1  Do you know the status of this PR? We are building control planes for LoRa and require this change. I'm unsure if the original author is still working on it. If the PR is stale, I can take over and continue the work.
